pkg/controller/controller.go

```diff
@@ -549,20 +749,72 @@ func (s *Service) DeleteVolume(ctx context.Context, req *csi.DeleteVolumeRequest
 			}
 		}
 
+		// used to restore session state after removing the volume
+		restoreMetroSession := func() error { return nil }
+
 		if len(vgs.VolumeGroup) != 0 {
 			// Remove volume from volume group
+			// Metro replication needs to be paused to remove the volume from the group.
+			var metroSessionID string
+			if isMetro {
+				isMetroVolumeGroup = true
+				metroSession, err := arr.Client.GetReplicationSessionByID(ctx, vgs.VolumeGroup[0].MetroReplicationSessionID)
+				if err != nil {
+					return nil, status.Errorf(codes.Internal, "unable to get metro session %s: %s", vgs.VolumeGroup[0].MetroReplicationSessionID, err.Error())
+				}
+
+				metroSessionID = metroSession.ID
+
+				// replication session should be in 'OK' or 'paused' state to continue
+				if metroSession.State != gopowerstore.RsStateOk && metroSession.State != gopowerstore.RsStatePaused {
+					return nil, status.Errorf(codes.FailedPrecondition,
+						"failed to delete volume %s because the metro replication session is in %s state", id, metroSession.State)
+				}
+				if metroSession.State == gopowerstore.RsStateOk {
+					// pause the replication session
+					_, err = arr.Client.ExecuteActionOnReplicationSession(ctx, metroSessionID, gopowerstore.RsActionPause, nil)
+					if err != nil {
+						return nil, status.Errorf(codes.Internal,
+							"failed to delete volume %s because the replication session could not be paused", id)
+					}
+
+					restoreMetroSession = func() error {
+						_, err = arr.Client.ExecuteActionOnReplicationSession(ctx, metroSessionID, gopowerstore.RsActionResume, nil)
+						if err != nil {
+							return status.Errorf(codes.Internal,
+								"failed to delete volume %s because the replication session could not be resumed", id)
+						}
+						return nil
+					}
+				}
+			}
+
 			// TODO: If volume has multiple volume group then how we should find ours?
 			// TODO: Maybe adding volumegroup id/name to volume id can help?
 			_, err := arr.GetClient().RemoveMembersFromVolumeGroup(ctx, &gopowerstore.VolumeGroupMembers{VolumeIDs: []string{id}}, vgs.VolumeGroup[0].ID)
 			if err != nil {
+				// try to restore metro replication before returning
+				if isMetro {
+					log.Info("could not remove volume from the volume group. Attempting to restore metro replication to its previous state.")
+					if e := restoreMetroSession(); e != nil {
```

Comment:
I don't think we need to resume here. DeleteVolume() will be called again and again until it can succeed. Resuming the session in failing case will only delays it.
So we can make the pause step idempotent. Please check.

Reply:
Ah yep that makes sense.